src/ds/address.h

```diff
@@ -39,4 +42,80 @@ namespace snmalloc
   {
     return reinterpret_cast<T*>(address);
   }
+
+  /**
+   * Test if a pointer is aligned to a given size, which must be a power of
+   * two.
+   */
+  template<size_t alignment>
+  static inline bool is_aligned_block(void* p, size_t size)
+  {
+    assert(bits::next_pow2(alignment) == alignment);
+
+    return ((static_cast<size_t>(address_cast(p)) | size) & (alignment - 1)) ==
+      0;
+  }
+
+  /**
+   * Align a pointer down to a statically specified granularity, which must be a
+   * power of two.
+   */
+  template<size_t granule, typename T = void>
+  inline T* pointer_align_down(void* p)
+  {
+    static_assert(granule > 0);
+    static_assert((granule & (granule - 1)) == 0);
+#if __has_builtin(__builtin_align_down)
+    return reinterpret_cast<T*>(__builtin_align_down(p, granule));
+#else
+    return reinterpret_cast<T*>(
+      bits::align_down(reinterpret_cast<uintptr_t>(p), granule));
+#endif
+  }
+
+  /**
+   * Align a pointer up to a statically specified granularity, which must be a
+   * power of two.
+   */
+  template<size_t granule, typename T = void>
+  inline T* pointer_align_up(void* p)
+  {
+    static_assert(granule > 0);
+    static_assert((granule & (granule - 1)) == 0);
+#if __has_builtin(__builtin_align_up)
+    return reinterpret_cast<T*>(__builtin_align_up(p, granule));
+#else
+    return reinterpret_cast<T*>(
```

Comment:
`address_cast` and `pointer_cast` might be clearer. I find `reinterpret_cast` horrible to read.